packages/repo/src/repo.ts

```diff
@@ -131,47 +134,50 @@ export class Repo extends ReadableRepo {
       }
     }
 
-    const unstoredData = await data.getUnstoredBlocks()
-    commitBlocks.addMap(unstoredData.blocks)
-
-    // ensure we're not missing any blocks that were removed and then readded in this commit
+    const dataCid = await data.getPointer()
     const diff = await DataDiff.of(data, this.data)
-    const found = commitBlocks.getMany(diff.newCidList())
-    if (found.missing.length > 0) {
-      const fromStorage = await this.storage.getBlocks(found.missing)
-      if (fromStorage.missing.length > 0) {
-        // this shouldn't ever happen
-        throw new Error(
-          'Could not find block for commit in Datastore or storage',
-        )
-      }
-      commitBlocks.addMap(fromStorage.blocks)
+    const newBlocks = diff.newMstBlocks
+    const removedCids = diff.removedCids
+
+    const addedLeaves = leaves.getMany(diff.newLeafCids.toList())
+    if (addedLeaves.missing.length > 0) {
+      throw new Error(`Missing leaf blocks: :${addedLeaves.missing}`)
     }
+    newBlocks.addMap(addedLeaves.blocks)
 
     const rev = TID.nextStr(this.commit.rev)
     const commit = await util.signCommit(
       {
         did: this.did,
-        version: 2,
-        prev: this.cid,
-        data: unstoredData.root,
+        version: 3,
         rev,
+        prev: this.cid,
+        data: dataCid,
       },
       keypair,
     )
-    const commitCid = await commitBlocks.add(commit)
+    const commitCid = await newBlocks.add(commit)
+
+    // ensure the commit cid actually changed
+    if (commitCid.equals(this.cid)) {
+      newBlocks.delete(commitCid)
+    } else {
+      removedCids.add(this.cid)
+    }
```

Comment:
yeah the commit CID would necessarily change because it contained the `prev` commit CID in the hash

but also we're just being more exact about block tracking in a way that we weren't previously